autoscheduler/scraper/banner_requests.py

```diff
@@ -0,0 +1,184 @@
+
+class BannerRequests():
+    """ Handles basic banner requests """
+
+    def __init__(self):
+        base_url = 'compassxe-ssb.tamu.edu'
+
+        self.depts_url = ('https://%s/StudentRegistrationSsb/ssb/classSearch/get_subject?'
+                          'dataType=json&offset=1&term={term}&max={max}' % base_url)
+
+        self.course_search_url = ('https://%s/StudentRegistrationSsb/ssb/searchResults/'
+                                  'searchResults/?pageOffset=0&sortDirection=asc&'
+                                  'sortColumn=subjectDescription&txt_subject={subject}&'
+                                  'txt_term={term}&uniqueSessionId={uniqueSessionId}&'
+                                  'pageMaxSize={num_courses}' % base_url)
+
+        self.create_session_url = ('https://%s/StudentRegistrationSsb/ssb/term/search?'
+                                   'mode=search&dataType=json' % base_url)
+
+        self.reset_search_url = ('https://%s/StudentRegistrationSsb/ssb/classSearch/'
+                                 'resetDataForm') % base_url
+
+    async def create_session(self, session: ClientSession, term: str) -> str:
+        """ Begins the session and validates the session_id
+            Must be called in order to search for courses
+        """
+
+        session_id = generate_session_id()
+
+        data = {
+            'uniqueSessionId': session_id,
+            'term': term,
+        }
+
+        await session.post(self.create_session_url, data=data)
+
+        return session
+
+    async def get_courses(self, session: ClientSession, session_id: str, dept: str, 
+                          term: str, amount: int) -> List[Dict]:
+        """ Retrieves all of the courses for a given department
+
+            dept: Department, a four letter string, such as CSCE
+            amount: max amount of courses to retrieve
+        """
+
+        data = {
+            'uniqueSessionId': session_id,
+            'term': term,
+            'subject': dept,
+            'num_courses': amount,
+        }
+
+        url = self.course_search_url.format(**data)
+
+        async with session.get(url) as response:
+            json = await response.json()
+
+        await self.reset_search(session)
+
+        data = json['data']
+
+        return data
+
+    async def get_departments(self, term: str, amount: int = 300) -> List[Dict]:
+        """ Retrieves all of the departments for the given term
+
+            Retrieving departments doesn't require an active session(nor session id),
+            so no need to call create_session(...)
+        """
+
+        data = {
+            'term': term,
+            'max': amount,
+        }
+
+        url = self.depts_url.format(**data)
+
+        response = requests.get(url) # Blocking, stops the thread until retrieves response
+
+        depts = response.json() # Also blocking
+
+        return depts
+
+    async def search(self, depts: List[str], term: str, amount: int = 750):

```

Comment:
If my understanding is correct (that search will synchronously call `get_courses` with each department in depts), I don't see a situation where doing this will be faster than making a thread for each department individually, so this function should probably just take in a single department.